zmq/backend/cython/socket.pyx

```diff
@@ -208,30 +208,109 @@ cdef inline int _proxy_step(void *handle_from, void *handle_to, int max_loops) e
     # If zmq_msg_init_* fails we must not call zmq_msg_close (Bus Error)
-    rc = zmq_msg_init_size(&data, msg_c_len)
+    rc = zmq_msg_init_size(&data.frame, data.msg_c_len)
 
     _check_rc(rc)
 
     with nogil:
-        memcpy(zmq_msg_data(&data), msg_c, zmq_msg_size(&data))
-        rc = zmq_msg_send(&data, handle, flags)
-        rc2 = zmq_msg_close(&data)
+        memcpy(zmq_msg_data(&data.frame), data.msg_c, zmq_msg_size(&data.frame))
+        rc = zmq_msg_send(&data.frame, handle, flags)
+        rc2 = zmq_msg_close(&data.frame)
     _check_rc(rc)
     _check_rc(rc2)
 
+cdef inline object _send_multipart_copy(void *handle, object msg_parts, int flags=0):
+    """Send a multipart message on this socket by copying its content."""
+    cdef int rc, rc2, rc3
+    cdef Py_ssize_t num_parts, base_part, batch_parts, partno
+    cdef int part_flags
+    cdef bint last_batch
+    cdef zmq_part_data_t data[32]
+    cdef object obj
+
+    num_parts = len(msg_parts)
+    base_part = 0
+
+    while base_part < num_parts:
+        # copy to c array:
+        batch_parts = num_parts - base_part
+        if batch_parts > 32:
+            batch_parts = 32
+            last_batch = False
+        else:
+            last_batch = True
+
+        # Copy the msg before sending. This avoids any complications with
+        # the GIL, etc.
+        # If zmq_msg_init_* fails we must not call zmq_msg_close (Bus Error)
+
+        for partno in range(batch_parts):
+            obj = msg_parts[base_part + partno]
+            if isinstance(obj, Frame):
+                obj = obj.buffer
+            zmq_part_data_init(&data[partno])
+            zmq_part_data_asbuffer_r(&data[partno], obj)
+
+        for partno in range(batch_parts):
+            rc = zmq_msg_init_size(&data[partno].frame, data[partno].msg_c_len)
+            if rc < 0:
+                # Close already initialized messages and break the loop
+                for partno in range(partno):
+                    zmq_msg_close(&data[partno].frame)
+                break
+
+        _check_rc(rc)
+
+        with nogil:
+            # Copy and send message parts
+            for partno in range(batch_parts):
+                memcpy(zmq_msg_data(&data[partno].frame), data[partno].msg_c, zmq_msg_size(&data[partno].frame))
+                if last_batch and partno >= (batch_parts - 1):
+                    # Last part, uses given flags
+                    part_flags = flags
+                else:
+                    part_flags = flags | ZMQ_SNDMORE
+                rc = zmq_msg_send(&data[partno].frame, handle, part_flags)
+                if rc < 0:
+                    break
+
+            rc2 = 0
+            for partno in range(batch_parts):
+                rc3 = zmq_msg_close(&data[partno].frame)
+                if rc3 < 0 and rc2 == 0:
+                    rc2 = rc3
+            if rc2 == 0:
+                rc2 = rc3
+
+        _check_rc(rc)
+        _check_rc(rc2)
+        base_part += batch_parts
+
```

Comment:
There doesn't appear to be any return value for this function.

Reply:
Returns `None`, as all python methods that don't return anything explicit. Or `NULL` if they raise an error.